modules/bazel_skylib/1.0.3/source.json

```diff
@@ -1,4 +1,5 @@
 {
-    "integrity": "sha256-HFMTdqx+WhgOAjeTiiU23gxU2T9cJ4Y0gY4O/JUt1Ww=",
-    "url": "https://github.com/bazelbuild/bazel-skylib/releases/download/1.0.3/bazel-skylib-1.0.3.tar.gz"
```

Comment:
@fmeum Do you know if url under `https://github.com/<org>/<repo>/releases/download` is deterministic or not?

Reply:
It feels strange that we cannot depend on them, also the content may actually be different that the tagged archive, for example `strip_prefix` isn't needed here.

Reply:
> @fmeum Do you know if url under `https://github.com/<org>/<repo>/releases/download` is deterministic or not?

Yep, these were explicitly claimed to be stable.

Reply:
Thanks, then @hobofan there is no need to replace those URLs.

Reply:
Given that everything under `/release/download` is user-uploaded, GH would have to actively have to temper with them for them not to be stable.

We could adjust the logic to have 3 states:
- `stable` (serves a consistent artifact, except for active interference by the maintainer like deleting the release or uploading a new artifact under the same name)
- `unstable` (may not serve the same artifact on consecutive requests; e.g. as it is dynamically produced in a non-deterministic way) - (bikeshedding: `volatile` might be a more appropriate naming 🙃 )
- `unknown` (anything else with unknown semantics)

We could then error for `unstable` URLs (as we can suggest a better `stable` alternative), and warn for `unknown` URLs.

Following that:
- Everything under `https://github.com/<org>/<repo>/releases/download` -> `stable`
- `https://github.com/<org>/<repo>/archive/refs/tags/<tag>.{tar.gz|zip}` -> `stable`
- Everything else under `https://github.com` -> `unstable`
- All other URLs: `unknown`

What do you think? I'd rather do it properly than too hastily, and than this could also serve as the basis for https://github.com/bazelbuild/bazel/issues/15128 in Bazel proper.

Reply:
Sounds good! 
However, we still have to think about how to deal with `https://github.com/<org>/<repo>/archive/$commit`. Maintaining a release on a fork sounds like too much extra work. My preference is also to give a warning on this, because I think in most case they are going to be stable, we also automatically mirror the archive to a bazel mirror. But of course, we should try to convince the original repo author to have releases or at least version tags, the current checked in version is only made up.